keel-tagging-plugin/src/main/kotlin/com/netflix/spinnaker/keel/tagging/ResourceTagger.kt

```diff
@@ -0,0 +1,203 @@
+    if (enabled) {
+      publisher.publishEvent(ScheduledResourceTaggingCheckStarting)
+      log.debug("Starting scheduled resource tagging…")
+      syncAccounts()
+      // todo emjburns: maybe one instance shouldn't do this whole thing
+      resourceRepository
+        .allResources { resourceHeader ->
+          if (resourceHeader.apiVersion != SPINNAKER_API_V1.subApi("tag")) {
+            try {
+              val tagResource = resourceRepository.get(resourceHeader.name.toTagSpecName(), KeelTagSpec::class.java)
+              if (tagResource.spec.tagState is TagNotDesired) {
+                // if it's in the resource repository, we want it to be tagged.
+                persistTagState(resourceHeader.name.generateKeelTagSpec())
+              }
+            } catch (e: NoSuchResourceException) {
+              persistTagState(resourceHeader.name.generateKeelTagSpec())
+            }
+          }
+        }
+      log.debug("Scheduled tagging complete")
+    } else {
+      log.debug("Scheduled tagging disabled")
+    }
+  }
+
+  // todo emjburns: should there be a catchup job for deletes?
+  @EventListener(DeleteEvent::class)
+  fun onDeleteEvent(event: DeleteEvent) {
+    log.debug("Persisting no tag desired for resource {} because it is no longer managed", event.resourceName.toString())
+    val spec = KeelTagSpec(
+      keelId = event.resourceName.toString(),
+      entityRef = event.resourceName.toEntityRef(),
+      tagState = TagNotDesired()
+    )
+    persistTagState(spec)
+  }
+
+  private fun persistTagState(spec: KeelTagSpec) {
+    log.debug("Persisting tag desired state for resource {}", spec.keelId)
+    val submitted = spec.toSubmittedResource()
+    val name = ResourceName(spec.generateTagNameFromKeelId())
+
+    if (tagExists(name)) {
+      resourcePersister.update(name, submitted)
+    } else {
+      resourcePersister.create(submitted)
+    }
+  }
+
+  private fun tagExists(tagResourceName: ResourceName): Boolean {
+    try {
+      resourceRepository.get(tagResourceName, KeelTagSpec::class.java)
+      return true
+    } catch (e: NoSuchResourceException) {
+      return false
+    }
+  }
+
+  private fun KeelTagSpec.generateTagNameFromKeelId() = "tag:keel-tag:$keelId"
+
+  private fun KeelTagSpec.toSubmittedResource() =
+    SubmittedResource(
+      apiVersion = SPINNAKER_API_V1.subApi("tag"),
+      kind = "keel-tag",
+      spec = this
+    ) as SubmittedResource<Any>
+
+  private fun ResourceName.generateKeelTagSpec() =
+    KeelTagSpec(
+      toString(),
+      toEntityRef(),
+      generateTagDesired()
+    )
+
+  fun ResourceName.generateTagDesired() =
+    TagDesired(tag = EntityTag(
+      value = TagValue(
+        message = KEEL_TAG_MESSAGE,
+        keelResourceId = toString(),
+        type = "notice"
+      ),
+      namespace = KEEL_TAG_NAMESPACE,
+      valueType = "object",
+      category = "notice",
+      name = KEEL_TAG_NAME
+    )
+    )
+
+  private fun syncAccounts() {
```

Comment:
Could you just do this with a different `@Schedule` annotation? It would make it easier to test.

Reply:
Yep, changing